src/api/root/handler.js

```diff
@@ -0,0 +1,183 @@
+
+const packageJson = require('../../../package.json')
+const Db = require('../../lib/db')
+const Util = require('../../handlers/lib/utility')
+const Kafka = require('../../handlers/lib/kafka/index')
+const Enum = require('../../lib/enum')
+const TransferEventType = Enum.transferEventType
+const TransferEventAction = Enum.transferEventAction
+
+const statusEnum = {
+  OK: 'OK',
+  DOWN: 'DOWN'
+}
+
+const serviceName = {
+  datastore: 'datastore',
+  broker: 'broker',
+  sidecar: 'sidecar',
+  cache: 'cache'
+}
+
+/**
+ * getHealth
+ *
+ * Gets the health status of the service.
+ *
+ *
+ * @example:
+ * returns the health of related sub-services:
+ * {
+ *   "status": "OK",
+ *   "uptime": 0,
+ *   "started": "2019-05-31T05:09:25.409Z",
+ *   "versionNumber": "5.2.3",
+ *   "services": [
+ *     {
+ *       "name": "datastore",
+ *       "status": "OK"
+ *     }
+ *   ]
+ * }
+ *
+ *
+ * @param {*} request - the request object
+ * @param {*} h       - the handler object
+ */
+const getHealth = async function (request, h) {
+  // Default values
+  let statusCode = 200
+  let status = statusEnum.OK
+  let isHealthy = true
+  let services = []
+  let subServices = {}
+
+  const uptime = process.uptime() // in seconds by default
+  const startTimeDate = new Date(Date.now() - uptime)
+  const startTime = startTimeDate.toISOString()
+  const versionNumber = packageJson.version
+
+  try {
+    services = [
+      await getSubServiceHealth(serviceName.datastore),
+      await getSubServiceHealth(serviceName.broker)
+    ]
+    isHealthy = evaluateServiceHealth(services)
+    subServices = {
+      services
+    }
+  } catch (err) {
+    isHealthy = false
+  }
+
+  if (!isHealthy) {
+    statusCode = 502
+    status = statusEnum.DOWN
+  }
+
+  const response = {
+    status,
+    uptime,
+    startTime,
+    versionNumber,
+    ...subServices
+  }
+  return h.response(response).code(statusCode)
+}
+
+/**
+ * getSubServiceHealth
+ *
+ * Gets the health status of a sub-service. Takes advantage of existing
+ * connections to the service (doesn't implement a ping or )
+ */
+const getSubServiceHealth = async function (serviceName) {
+  switch (serviceName) {
+    case 'datastore': {
+      const tables = await Db._listTables()
+
+      return {
+        name: serviceName,
+        status: tables.length > 0 ? statusEnum.OK : statusEnum.DOWN
+      }
+    }
+    case 'broker': {
+      /*
+        Ensure that all topics for consumers and producers are available
+        Otherwise consider Kafka DOWN
+      */
+      const consumerTopics = [
+        Util.transformGeneralTopicName(TransferEventType.TRANSFER, TransferEventAction.PREPARE),
+        Util.transformGeneralTopicName(TransferEventType.TRANSFER, TransferEventAction.POSITION),
+        Util.transformGeneralTopicName(TransferEventType.TRANSFER, TransferEventAction.FULFIL)
+      ]
+
+      let status = statusEnum.OK
+      try {
+        await P.all(consumerTopics.map(t => Kafka.Consumer.getConsumer(t)))
```

Comment:
The Kafka streaming lib (https://github.com/mojaloop/central-services-stream) does not expose any method to check connectivity, however, the node-rdkafka library does with "isConnected()". It may be better to update the library appropriately to expose a similar method to support the health check purpose.

The approach above will work on the initial setup of the service, but we have seen times where the consumers are disconnected. This approach would still report "ok" in that scenario.

Reply:
Sure. I'll take a look at updating that library.

Reply:
> Sure. I'll take a look at updating that library.
> 
> We might also have to handle a case where there are different hosts for each of the topics as well - as I've noticed that the hostname of the broker can be set per topic in the config file.

Totally agree with you there, but implementing the "isConnected()" method should cater for this since you will grab all the consumers and just run the "isConnected" method to ensure that all is well.